### ISSUE TITLE
Write proposal for Permissions interface

### ISSUE THREAD

**Issue by mistermoe** [decentralized-identity/identity-hub#126]

**Comment by mistermoe** [comment]
# Permission Representation
Permissions are represented as JWTs and adhere to the [UCAN spec](https://github.com/ucan-wg/spec)

## JWT Structure
### Header

| Field | Type     | Description                    | Required |
| ----- | -------- | ------------------------------ | -------- |
| `alg` | `String` | Signature algorithm            | Yes      |
| `typ` | `String` | Type (MUST be `"JWT"`)         | Yes      |

### Payload

| Field | Type       | Description                                      | Required |
| ----- | ---------- | ------------------------------------------------ | -------- |
| [`iss`](https://github.com/ucan-wg/spec#321-principals) | `String`   | Issuer DID (sender)                              | Yes      |
| [`aud`](https://github.com/ucan-wg/spec#321-principals) | `String`   | Audience DID (receiver)                          | Yes      |
| [`nbf`](https://github.com/ucan-wg/spec#322-time-bounds) | `Number`   | Not Before UTC Unix Timestamp (valid from)       | No       |
| [`exp`](https://github.com/ucan-wg/spec#322-time-bounds) | `Number`   | Expiration UTC Unix Timestamp (valid until)      | Yes      |
| [`nnc`](https://github.com/ucan-wg/spec#323-nonce) | `String`   | Nonce                                            | No       |
| [`prf`](https://github.com/ucan-wg/spec#325-proof-of-delegation) | `String[]` | Proof of delegation (witnesses are nested UCANs) | Yes      |
| [`att`](https://github.com/ucan-wg/spec#325-attenuations) | `Json[]`   | Attenuations. An array of [Capability objects](#Capability) | Yes      |
| [`fct`](https://github.com/ucan-wg/spec#324-facts) | `Json[]`   |  | Facts (asserted, signed data)      |




## Capability
- A capability (often referred to as "caveat" in previous OCAP art) is used to attenuate the scope of a permission and adheres to the capability format defined in the [UCAN spec](https://github.com/ucan-wg/spec#325-attenuations). 
- A capability is the association of an [ability](https://github.com/ucan-wg/spec#3242-ability) to its [resource](https://github.com/ucan-wg/spec#3241-resource-pointer)
- A capability **MUST** be represented as a JSON object that contains the following fields:


| Field          | Type        | Description               | Required |
| -------------- | ----------- | ------------------------- | -------- |
| `with`         | `string`    | [See here](#Resource)     | Y        |
| `can`          | `Object`    | [See here](#Ability)      | Y        |
| `constraints`  | `Object`    | [See here](#Constraints)  | Y        |


### Resource
- the [resource](https://github.com/ucan-wg/spec#3241-resource-pointer) **MUST** be the DID of the issuer

### Ability 
- an [ability](https://github.com/ucan-wg/spec#3242-ability) **MUST** be represented as a JSON object that contains the following fields:

| Field          | Required (Y/N) |
| --------       | -------- |
| `method`       | Y     |
| `schema`       | N     |
| `objectId`     | N     |
| `description`  | N     |

- an ability **MUST** contain the `method` field. all additional fields are optional and are used to further limit the scope of what the capability permits
- the value `*` is reserved to represent all possible values for a given field
    - e.g. `method: *` can be interpreted as "All current and future available methods"
- the fields within a capability are considered hierarchical. The hierarchy is as follows: 
    ```
    method > schema > objectId
    ```
    - An ability that includes a specific field **MUST** include all cardinal fields.
        - e.g. a capability that includes `objectId` **MUST** include `schema` and `method`
    - any omitted subsidiary fields within a capability are implicitly represented as `field: *`
        - e.g. `{ "method": "CollectionsWrite" }` is equivalent to `{ "method": "CollectionsWrite", schema: "*", "objectId": "*" }`

### Constraints 
- Constraints are an optional declaration of attestation and encryption requirements (e.g. Data MUST be encrypted). 
- Constraints **MUST** be represented as a JSON object that contains the following fields:

| Field | Type       | Description                                        | Required |
| ----- | ---------- | ------------------------------------------------   | -------- |
| `encryption`    | `int`       | Whether associated data should be encrypted | N     |
| `attestation`   | `int`       | Whether associated data should be signed    | N     |
| `sharedAccess`  | `boolean`   | Whether the grantee has access to the objects of other entities | N     |

- the value of `encryption` **MUST** be an unsigned integer. where: 

    | Code | Description |
    | -------- | -------- |
    | `0`     | MUST NOT Encrypt     |
    | `1`     | JWE     |

    - Any value `>0` implies that any associated `data` **MUST** be encrypted as described in the [Encrypted Data](https://identity.foundation/identity-hub/spec/#encrypted-data) section

- the value of `attestation` **MUST** be an unsigned integer. where: 

    | Code | Description |
    | -------- | -------- |
    | `0`     | MUST NOT SIGN     |
    | `1`     | JWS     |

    - Any value `>0` implies that any associated data MUST be attested as described in the [Signed Data](https://identity.foundation/identity-hub/spec/#signed-data) section



## Root Capability
**TODO**: Fill out


# Permission Invocation
Invoking a permission is done by signing over the Permission JWT. The resulting JWS is sent along in a message as the value of `authorization`

# Interface

## `PermissionsRequest`
Permissions can be requested using the `PermissionsRequest` method
```json
{
    "descriptor": {
        "method": "PermissionsRequest",
        "objectId": "f45wve-5b56v5w-5657b4e-56gqf35v",
        "audience": "did:ex:alice",
        "ability": {
          "method": "CollectionsWrite",
          "schema": "https://schema.org/MusicPlaylist"
        }
    }
}
```

- A `PermissionsRequest` message **MUST** contain the following fields:


| Field | Type | Required (Y/N) | Description |
| -------- | ------- | -------- | -------- |
| `descriptor.method`    | `String` | Y     | MUST be `PermissionsRequest`     |
| `descriptor.audience`    | `String` | Y     | DID of the party requesting permission     |
| `descriptor.ability`   | `Object`  | Y     | The desired ability     |
| `descriptor.objectId`    | `String` | Y     | A UUID-v4     |
| `attestation`    | ? | ?     | ?     |


## `PermissionsGrant`

Permissions can be granted using the `PermissionsGrant` method
```json
{
    "descriptor": {
      "method": "PermissionsGrant",
      "objectId": "f45wve-5b56v5w-5657b4e-56gqf35v",
      "grantedFor": "78t9tnb-e65vses-ver678fuc-56gqf35v",
      "cid": "e34455ev6v365eb7r8n9mnbesv5e6be7rn879m",
      "dataFormat": "application/json"
    },
    "data": {
      "payload": {
        "iss": "did:ex:alice",
        "aud": "did:ex:bob",
        "att": [{
          "with": "did:ex:alice", // Is this right ???
          "can": {
            "method": "CollectionsWrite",
            "schema": "https://schema.org/MusicPlaylist"
          },
          "constraints": {
            "sharedAccess": true
          }
        }]
      },
      "signature": "fw547v63bo5687wvwbcqp349vwo876uc3q..."
    } 
}
```

- A `PermissionsGrant` message **MUST** contain the following fields:

| Field | Type | Required (Y/N) | Description |
| -------- | ------- | -------- | -------- |
| `descriptor.objectId`   | `String`  | Y     | A UUID-v4     |
| `descriptor.method`     | `String`  | Y     | MUST be `PermissionsGrant`  |
| `descriptor.dataFormat` | `String`  | Y     | MUST be `application/json`    |
| `descriptor.cid`        | `String`  | Y     | The CID of the data that represents the granted UCAN JWT   |
| `data`                  | `Object`  | Y     | The granted UCAN JWT    |
| `attestation`           | `Object`  | Y     | The signed-over grant    |

## `PermissionsRevoke`


Permissions can be revoked using the `PermissionsRevoke` method
```json
{
    "descriptor": {
      "method": "PermissionsRevoke"
      "objectId": "f45wve-5b56v5w-5657b4e-56gqf35v"
    }
}
```

- A `PermissionsRevoke` message **MUST** contain the following fields:

| Field | Type | Required (Y/N) | Description |
| -------- | ------- | -------- | -------- |
| `descriptor.objectId`   | `String`  | Y     | A UUID-v4     |
| `descriptor.method`     | `String`  | Y     | MUST be `PermissionsRevoke`  |

# Discussion Topics

**Comment by mistermoe** [comment]
Resolved by #129
